Added a method to set and get additional spacing between characters.

Foaly opened the pull request:
**SFML Tasks**

* [x] Check for forgotten discussion points
* [x] Review code

----

This is my code, that I promised in [this forum thread](http://en.sfml-dev.org/forums/index.php?topic=18635.0).
It is pretty straight forward. Here is some example code to test with:

``` cpp
#include <SFML/Graphics.hpp>
#include <iostream>

int main()
{
    sf::RenderWindow window(sf::VideoMode(800, 600), "Character Spacing Example");
    window.setFramerateLimit(60);

    // load font
    sf::Font font;
    if(!font.loadFromFile("BilboSwashCaps-Regular.otf"))
    {
        std::cerr << "Failed to load font!" << std::endl;
        return 1;
    }

    // setup the texts
    sf::Text normalText("SFML rocks! 123", font);
    normalText.setCharacterSize(100);
    normalText.setPosition(30, 10);
    normalText.setStyle(sf::Text::Style::Bold);

    sf::Text smallText("SFML rocks! 123", font);
    smallText.setCharacterSize(30);
    smallText.setPosition(30, 150);
    smallText.setStyle(sf::Text::Style::Bold);

    sf::Text stretchedText("SFML rocks! 123", font);
    stretchedText.setCharacterSize(100);
    stretchedText.setPosition(30, 200);
    stretchedText.setStyle(sf::Text::Style::Bold);
    stretchedText.setLetterSpacing(6.f);

    sf::Text smallStretchedText("SFML rocks! 123", font);
    smallStretchedText.setCharacterSize(30);
    smallStretchedText.setPosition(30, 350);
    smallStretchedText.setStyle(sf::Text::Style::Bold);
    smallStretchedText.setLetterSpacing(1.f);


    // run the program as long as the window is open
    while (window.isOpen())
    {
        // check all the windows events
        sf::Event event;
        while (window.pollEvent(event))
        {
            // "close requested" event: we close the window
            if (event.type == sf::Event::Closed)
                window.close();

            else if (event.type == sf::Event::KeyReleased) {
                // close if escape key was pressed
                if (event.key.code == sf::Keyboard::Escape) {
                    window.close();
                }
            }
        }

        // clear the window to black
        window.clear();

        // draw text
        window.draw(normalText);
        window.draw(smallText);
        window.draw(stretchedText);
        window.draw(smallStretchedText);

        // display the windows content
        window.display();
    }
    return 0;
}
```

![letter-spacing](https://cloud.githubusercontent.com/assets/1014762/8810496/a9c169fc-2fef-11e5-93d9-7329b3846102.png)
The font Bilbo Swash Caps Regular is taken from [here](http://openfontlibrary.org/en/font/bilbo).

TankOs commented on include/SFML/Graphics/Text.hpp:
```diff
@@ -145,6 +145,21 @@ class SFML_GRAPHICS_API Text : public Drawable, public Transformable
     void setCharacterSize(unsigned int size);
 
     ////////////////////////////////////////////////////////////
+    /// \brief Set the addition character spacing
```
"additional"?

TankOs commented:
Except one doc string, this looks good to me. :+1: :cat2:

MarioLiebisch commented:
I'm not 100% sure. How about introducing scaling of the horizontal (and vertical) offsets rather than an absolute offset? Or in other words, make character spacing and line spacing configurable.

As far as I know, stuff like  s p a c e d   o u t   t e x t would usually scale with the actual font/glyph's default offsets.

Foaly commented:
Thanks, I updated the doc string!

I also already thought about line spacing. Although I personally didn't need it yet, it has been requested a couple times on the forum, for example [here](http://en.sfml-dev.org/forums/index.php?topic=9620.0). Laurent more or less agreed to it [here](http://en.sfml-dev.org/forums/index.php?topic=16451.0). I also think it is a basic typography tool and it would be a useful addition to `sf::Text`. What does everybody else think? If nobody has anything against it, I would send another commit implementing line spacing.

About the scaling factor vs. absolut offset: I took a quick look around and here is what others do. GIMP uses only absolut offsets for both letter and line spacing. In LibreOffice letter spacing can only be set using absolut offsets. Line spacing can be defined by both offsets and factors (double line, 1.5 line etc.). In CSS the `letter-spacing` property only takes absolut pixel values. The `line-height` property takes takes both offset and scaling factors.
So I think we should stick with absolut offsets for letter spacing, since it seems to be the standart way. Since it's a pixel offset, it also gives the user the most control, I'd say. For line-spacing I'd also suggest absolut offsets to be consistend and least confusing.

Another small thing: In [this thread](http://en.sfml-dev.org/forums/index.php?topic=16451.0) the name `setLetterSpacing()` was suggested. [Letter Spacing](https://en.wikipedia.org/wiki/Letter-spacing) seems to be the correct term for it. I used `setCharacterSpacing()` in analogy to `setCharacterSize()`. What are your opinions on the name? I personally think `setLetterSpacing()` is actually more self-explanatory.

MarioLiebisch commented:
*Letter spacing* might be the traditional term, but modern typefaces have far more characters than just letters (like dingbats). So I'd stick to SFML's *character*.

Scaling vs. offsets: If offsets are more common, stick with these. :)

mantognini commented:
Doesn't this implementation add one-too-many `m_characterSpacing` to the width of non-empty strings?

Bromeon commented:
> *Letter spacing* might be the traditional term, but modern typefaces have far more characters than just letters (like dingbats).

I don't think this is relevant. If there is a common term ([which is the case](https://en.wikipedia.org/wiki/Letter-spacing)), we should use it, not work our way around conventions.

MarioLiebisch commented:
Okay, if it's "official". :)

TankOs commented:
:+1: for both absolute offsets and line spacing. What do others say?

Foaly commented:
@mantognini: I don't think so. If I understand the implementation correct, the bounds are computed before the [advance is added](https://github.com/Foaly/SFML/blob/characterSpacing/src/SFML/Graphics/Text.cpp#L393). So for the last letter in a non-empty string the width is computed, then the advance is added, but then the loop is left and the bounds stay correct.

OK, so I guess we'll go with `setLetterSpacing()`. I'll change that. I'll see if I have time to implement line spacing later today, so we have something to discuss.

mantognini commented:
@foaly could you make sure that's the case? E.g. By checking that a one char text has always the same size regardless of the spacing. Just to be sure :)

Foaly commented:
@mantognini: I did a quick check. A text with one character returns exactly the same bound compiled with this branch and with current master.

Also I updated the commit. Changed the names and the documentation accordingly.

Foaly commented:
Alright, so I implemented line spacing. The implementation follows the letter spacing one closely. It's simply an offset, that is added onto the line spacing from the font.
Right now I broke some of the indentation, but it's about the implemenation for now. Also the more I look at `ensureGeometryUpdate()`, the more little things I'd like to change I see (const is not used consequently, some code dublication...). So if nobody minds I would like to refactor that method a bit. I won't be able to do it today, but probably on the weekend.

LaurentGomila commented:
> if nobody minds I would like to refactor that method a bit

Then do it in a separate PR please, don't mix unrelated changes.

zsbzsb commented:
>Also the more I look at ensureGeometryUpdate(), the more little things I'd like to change I see (const is not used consequently, some code dublication...).

Look at my text outline PR, I already refactored quite a bit.

Foaly commented:
Thanks! Yeah I had something similar in mind. A while back I submitted #442 which already incorporated some of those changes, but the refactoring was never added, because vtab got removed and the PR was closed.

I'll probably have time to submit the changes tonight.

@LaurentGomila I know making a separate pull request is good practice and I can do it, if you want. But are you sure in this case? I mean the changes are quiet small and related in my opinion (added new functionality -> refactored existing method). Also a separate pull request would mean merge conflicts. Your call though.

Bromeon commented:
@Foaly As a compromise, and to avoid unnecessary work with merge conflicts, you could make two separate commits: one for the refactoring, and one for the actual feature.

Foaly commented:
Alright here are my changes! I'd love to hear some feedback.

zsbzsb commented:
As I already refactored the code to add a line to an external method in #840 I'm not sure why you did it again.

Foaly commented:
If you take a closer look, you'll see that I the method has a different sigature and a slightly different implementation. I found it more expressiv this way.

Bromeon commented:
But @zsbzsb has got a point: now we have definitive merge conflicts and the merge order determines who is "lucky" and needn't fix his code. Since zsbzsb has opened his pull request already in March, I suggest we use his code... and if needed, you @Foaly adapt it correspondingly.

I don't think it makes a lot of sense to rewrite everything now. We should wait until his PR is merged, so that you can rebase on master.

Foaly commented:
I opened a pull request with that had this exact refactoring almost [two years ago](https://github.com/SFML/SFML/pull/460)... But anyway, if you want I can take that change out and give @zsbzsb some feedback on how to improve his code.
Maybe I'll find some time later today.

Bromeon commented:
Okay, I didn't know. In any case, we should agree on one refactoring, it needn't be zsbzsb's. 

Maybe Laurent was not so wrong with the suggestion to have a separate PR just for that :wink:
If we have such a thing, we can merge it, and both PRs can base on that code...

Foaly commented:
OK so before this pull request get cold(er): What should we do here? Possibilities:
 * I could remove the refactoring entirly
 * Somebody could merge one of the pull request and the other one adjusts his pull request accordingly (I don't mind)
 * We could both remove our refactoring, focus the PR's only on the feature and send a new PR containing only refactoring (only makes sense after the other two PR's are merged)

zsbzsb commented:
Here's my take on it. My PR directly affects those lines of code that add a line to the rendered text. In fact, without the refactoring my code would make the duplicated code even longer ([count how many times 'addLine' is called - 8 times up from just 4](https://github.com/SFML/SFML/pull/840/files)). Also my PR adds a second vertex array to the text class to draw the outline. That means the refactored code will already be outdated (it can't handle separate vertex arrays unless that is added with knowledge of the future text outlining PR). So if any refactoring is merged before text outlining then my changes (after rebasing) will just about replace all of the changes in the refactoring.

I'm not in charge of merging so whatever happens I will update PR as required, I just think its clear what should happen.

Foaly commented:
Thanks! I think the feature you are adding is awesome as well! I really like typography, so it's cool to see that `sf::Text` is actively being improved. 
I totaly agree. I also want to see new features getting merged rather than sitting around waiting and have two PRs block each other.

You made the point clear. Your code definitly effects the existing code more than mine does. So I agree your code should be merged first. I will remove my refactoring and adapt my code once your's is merged. I will add the other small improvments I made (const, variable naming, etc...) again afterwards.

One suggestion though. Take a look at my `appendLine()` method. [Here](https://github.com/SFML/SFML/pull/928/files#diff-d2db48e7e604a8901d8739c523d30981R419) and [here](https://github.com/SFML/SFML/pull/928/files#diff-4e2dd8abc6319bc102b837bdaaa70df7R352) I find the parameters and their naming a little more expressive and elegant. It also make the implementation a little cleaner. Also I used `roundf()` instead of `std::floor(... + 0.5f)`.

Now we just have to wait for your PR to get merged :smile:

LaurentGomila commented:
> Also I used roundf()

As far as I know, it's C++11.

Foaly commented:
Nope std::round() is c++11 roundf() is c++98.

LaurentGomila commented:
I thought it was C99, thus C++11.

Bromeon commented:
> Nope std::round() is c++11 roundf() is c++98.

That's not true, it's C99. It has **not** been adopted to C++11, even if some compilers support it. [[source]](http://en.cppreference.com/w/cpp/numeric/math/round)

Foaly commented:
OK sorry my bad. But even when it's C99, you can still include `<math.h>` and use `roundf()`.

Bromeon commented:
> OK sorry my bad. But even when it's C99, you can still include  <math.h>  and use  roundf() .

Not in SFML... We use the **C++98** standard, not C. You can use `std::floor(x + 0.5f)` however.

LaurentGomila commented:
> But even when it's C99, you can still include <math.h> and use roundf().

Not if you compile with a C++ compiler. <math.h> from the C++ (03) compiler and <math.h> from the C (99) compiler are different. These are different languages with different standard libraries.

Bromeon commented:
Exactly. To extend on that, C89/90 is *mostly** compatible to C++98/03, but starting from C99, the languages C and C++ diverge.

\___________
\* There are some language features which are allowed in C, but not C++, mostly regarding type safety.

Foaly commented:
Ok so since #840 is merged I will update the PR. I'll try to do it soon™

zsbzsb commented:
:+1:

Foaly commented:
Alrighty I rebased my code onto current master. The first two commits remained untouched. I of course had to redo the code style refactor commit, so a quick review would be nice. I tested the new state of the PR with the test code above (which I updated). It compiles and works as expected.
I also updated the documentation so it mentions character/line spacing and text outlines. And one example was using the deprecated setColor.
So everything should be ready for merging :smiley:

zsbzsb commented:
My 2 cents on the refactoring... The change so all of the local variables are marked as const (not that I care), but AFAIK that isn't consistent with the rest of the code base. Checking the [code style](http://www.sfml-dev.org/style.php) all I found was "```const``` is placed before the type whenever possible. Reference & or pointer * are glued to the type (no extra space)." I'm not sure how that applies here as just about no other function I have seen marks the local variables as const.

Also I'm not sure if the change from ```hspace``` to ```whitespaceWidth``` and ```vspace``` to ```lineSpacing``` makes the code more readable.

>I also updated the documentation so it mentions character/line spacing and text outlines. And one example was using the deprecated setColor.

Thanks, I missed that :smile:

Foaly commented:
Well I agree it is not constistent across the entire codebase. But you see it a lot especially in newer parts. It's a good practice to mark variables as const, unless you want to modify them. Especially in bigger functions it makes it easier to see what variables are not changed on first sight. And also its in the code style guide line.
Hmm I don't know I think the new variable names are more expressive. 

> .... deprecated setColor ...

My compiler warned me because of the new deprecated macro :+1:

Bromeon commented:
Generally, you should not modify the style of existing code unless it notably improves its quality or makes it adhere better to *our* style guide, especially not if that modification makes it diverge from code style in other parts of the library. I agree that some variables are now named better, but are e.g. `isBold` etc. consistent with existing SFML code? There's nothing saying that boolean flags need an "is" prefix, I'd rather think about naming `italic` more expressively ("italicShear", "italicRotation").

Also regarding `const` for local variables, there are different views. It undoubtedly makes the variable unmodifiable and can thus prevent errors (however, do those really appear frequently?). But, it makes the code a tiny bit less readable, and having every variable `const` distracts from more important parts and code where `const` is semantically required. Furthermore, it's inconsistent with parameters, which should then be const-qualified as well, and this would again move implementation details to the header when declaration and definition are the same (e.g. for templates). When the function is so big that it becomes unclear which variables are used how, a split can be considered.

The statement "const is placed before the type whenever possible" means that the keyword should appear on the left hand side of the type, not the right (`const T` instead of `T const`), not that you should add it whenever possible :wink: because that would be a really stupid rule. It's ambiguously written, I know, but the following code section definitely clarifies how it's meant.

Foaly commented:
Hmm I don't know I think it's a good practice to put `const` infront of variables that you don't want to modify. In my opinion it makes to code easier to read when you are not familiar with it. I did it for the new variables I added, so it seemed logical to updated the existing ones. I agree it's not consistent across the codebase, but like I said it found it in a lot of newer commits. I admit that I understood the code style guidline different though. `const` for parameter doesn't make sense (exept pointer and references) because the values are copied anyway.
But of course it's the teams choice. I can change it if you want.
I actually adapted the practice of naming bool variables `bool isSomething` from the SFML codebase. I think its awesome, because you understand on first read what state the variable represents. So yes I guess it is consistent with the existing code.
I really like the name `italicShear`! I will update the code tomorrow.

Foaly commented:
Ok I updated the code again.
'tis the question now: to const or not to const? :smile:

eXpl0it3r commented:
I'd say not to const. :wink:

As @Bromeon said, the function parameters are not const either and they are local variables as well.

In the beginning I thought as well that one should mark as much as possible as const to be "const correct", but I then learned that in most cases when it doesn't really matter if its const or not, it's more accepted to not make it const; nobody using the API cares if you change the local variable. Const correctness is mostly important regarding const functions and Rvalues.

Foaly commented:
I don't get the argument that the function parameters are not const... What functions are you talking about?! The parameters of the methods in the class are either POD or they are correctly using const! Making the POD const makes no sense, because they are copied anyway.

Of course nobody using the API is gonna care if local variables are const, but people who invest the time to look at the source to understand how things work or to improve things will have an easier time finding their way around. It is more obvious on first sight what is happening. You can boldy see what is being modified and what is constant. It makes the code more readable in my opinion. Also it assists the (library) programmer to prevent programming errors.

The only thing I can really see speaking against it is, that it is inconstinent across the code base. But why not progressivly change it to something more expressive? Again like I said quiet a lot of the newer commits use const local variables (like: 9c5c750e60f33aeb00ed07bf80810bd49b4f8947 or b3b48a507d9c941607bcd6b616c399687769bae2)

But I really don't want to fight about something this small and unimportant. I just don't understand your arguments... If you guys really want me to change it ill change it.

Bromeon commented:
A parameter is used like a local variable (same lifetime and visibility scope), thus the same arguments apply. We're talking about function definitions, not declarations.

And as I explained above, it's not necessarily more readable... It makes it easier to overlook semantically more important consts (like const-references/pointers). Maybe read that comment again.

Foaly commented:
I update it.

One more question: Would it be a good idea to caculate the `italicShear` instead of using a magic number? I mean doing something like this:
```cpp
float italicShear = (m_style & Italic) ? (12.f / 360.f * 2.f * PI) : 0.f; // 12 degrees italic shear
```
It would have the advantage of clearly documenting how this number is calculated in the code. I makes it easier to change. It uses the full precision of float (could be interesting for high zoom levels). Every compiler would evaluate this at compile time, so there is no overhead.

Bromeon commented:
Thanks for the updates. I think with the comment it's clear enough, the `= (12 / 360 * 2 * PI)` part could even be omitted (no need to explain how to convert degrees to radians), something like `// 12 degrees as radians` might even be more expressive.

You're right about it being easier to change, but will this really be done often? The whole expression would get quite big and needed to have a separate function, also because there's no Pi constant in standard C++... so I doubt it's worth the verbosity, as it takes 3 seconds to convert such a value online :wink:

Foaly commented:
It was just an idea. I update the code. I'm guessing using the whole float range (0.20943952f) is not necessary?

zsbzsb commented:
Honestly I don't care if the value is calculated at compile time (by writing in the formula) or just pre-calculated (as it is now). Either way won't make a difference speed wise - as for the extra precision? That probably wouldn't make a difference either due to floating point inaccuracy and because we are limited to rendering physical pixels.

TankOs commented:
:+1: for readability.

Foaly commented:
Well ok. If nobody cares than this is finally ready for merge now I would say.

TankOs commented:
:+1: :cat2:

zsbzsb commented:
Looks good :+1:

Foaly commented:
Bump!

Bromeon commented:
Looks good and works fine on Windows. :+1:

[Here](https://gist.github.com/Bromeon/d64b9e08f4a6c82f80f9) is the Gist I used to test the letter and line spacing.

Foaly commented:
Thanks for testing!
About the units. In a text editor scenario you might be more used to defining the spacing with a factor or even points, but I think in the context of games or multimedia applications pixels make more sense, because you often want pixel perfect control about how to position things. Thats at least what I think, can others think of other use cases?
Since you define a _additional_ offset with these setter functions 1.5 times the line spacing could be done like this: `text.setLineSpacing(font.getLineSpacing(text.getCharacterSize()) * 0.5f)`

eXpl0it3r commented:
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.

Bromeon commented:
It would be nice to have more opinions on these points, before the API is frozen and people regret it:

> **Bromeon:** is it intuitive if pixels are used as the spacing unit? Wouldn't people expect a factor (e.g. 1.5 times line spacing)? How difficult is it now to achieve a multiple of the default spacing (to leave every second line empty, for example)?

> **Foaly:** In a text editor scenario you might be more used to defining the spacing with a factor or even points, but I think in the context of games or multimedia applications pixels make more sense, because you often want pixel perfect control about how to position things. Thats at least what I think, can others think of other use cases?

eXpl0it3r commented:
> Wouldn't people expect a factor (e.g. 1.5 times line spacing)?

Maybe, but the question then is, can you get pixel perfect spacing just by using factors? How much do floats kill the precision?

Then again the API as is, uses floats for pixels, which makes me wonder what `1.3245f` pixels is? Meaning, if we go for pixels, shouldn't it be an integer type?

> How difficult is it now to achieve a multiple of the default spacing?

`text.setLineSpacing(font.getLineSpacing(text.getCharacterSize()) * 0.5f)`

Hmmm. Shouldn't the API name be setLineSpacingOffset then? Or if we stick with the name, shouldn't the default be `1.f * font.getLineSpacing()` instead of 0? I find it otherwise kind of misleading.

One could even consider renaming the font function to `getDefaultLineSpacing` or similar, but since it's part of the font, the API might be verbose enough already.

> (to leave every second line empty, for example)?

Unless you meant it differently, you're mixing things here. Line height and default line spacing are not the same thing, but I guess you mean "line" as in "line spacing"-line.

eXpl0it3r commented on include/SFML/Graphics/Text.hpp:
```diff
@@ -145,6 +145,36 @@ class SFML_GRAPHICS_API Text : public Drawable, public Transformable
     ////////////////////////////////////////////////////////////
+    /// \brief Set the additional line spacing offset
+    ///
+    /// The spacing between lines is defined by the font.
```
Maybe add "default" to all of these similar comments ,e.g.

`The default spacing between lines is defined by the font.`

Bromeon commented:
Thanks for your time to comment! You mention some really good points.

> Maybe, but the question then is, can you get pixel perfect spacing just by using factors? How much do floats kill the precision?

Is the loss of precision of a single multiplication relevant? The variables will hold numbers close to the desired value, but when rasterized with OpenGL, the effect will most probably not be visible.

> If so I think "pixel" might not be the right term for the docs.

Are we talking about *points* rather than pixels? Here's a [pixel vs points](http://graphicdesign.stackexchange.com/a/207) SE thread.

> Hmmm. Shouldn't the API name be setLineSpacingOffset then? Or if we stick with the name, shouldn't the default be `1.f * font.getLineSpacing()` instead of 0? I find it otherwise kind of misleading.

I totally agree.

The reason why I asked those questions is that line/character spacing **factors** are probably more intuitive than **offsets**. Many people are used to them from software like Microsoft Word or similar.

What we also should consider: when the character size increases, the absolute offset will remain the same, making the characters appear closer/tighter. Using a spacing factor on the other hand will preserve the appearance as the characters are varied in size. It probably depends on concrete UI needs, but I find the latter behavior more intuitive.

eXpl0it3r commented:
Any further comments or actions that should be taken?

Bromeon commented:
It would be especially interesting to hear other people's opinions on the points mentioned by the two of us. So far, I have the feeling that an API based on factors rather than offsets might be easier to use and more intuitive.

binary1248 commented:
I'm also for using relative factors instead of absolute values (pixels, points etc.). However... I also think we should "consult" freetype in order to compute the proper spacings to apply the factor to.

Adding a fixed amount between characters or lines will lead to a very "disharmonious" feeling when reading the text, since naturally the advance between characters depends on which 2 characters are being considered.

The spacing between lines isn't that critical, but I agree with @eXpl0it3r that we should start with a factor of 1 * default line spacing and work from there. There are already many other 1-based factors in use now, such as viewport and drawable scaling.

Foaly commented:
I think there is a misconception here. The way the implementation works right now, is that the setters specifiy an **additional** offset. Meaning that the spacing is aquired from freetype depending on the character size, just as you would expect, but then an additional pixel offset is added ontop of that. (the doumentation specifies that, but maybe the method name should be changed to something like setLine/LetterSpacingOffset() or setAdditionalLine/LetterSpacing() like eXpl0it3r suggested)

I wrote a [comment](https://github.com/SFML/SFML/pull/928#issuecomment-123696455) all the way back in july where I did a quick look around, what others do.
> About the scaling factor vs. absolut offset: I took a quick look around and here is what others do. GIMP uses only absolut offsets for both letter and line spacing. In LibreOffice letter spacing can only be set using absolut offsets. Line spacing can be defined by both offsets and factors (double line, 1.5 line etc.). In CSS the letter-spacing property only takes absolut pixel values. The line-height property takes takes both offset and scaling factors.
So I think we should stick with absolut offsets for letter spacing, since it seems to be the standart way. Since it's a pixel offset, it also gives the user the most control, I'd say. For line-spacing I'd also suggest absolut offsets to be consistend and least confusing.

I think my conclusion from back then is still valid. For letter spacing absolut offsets make more sense then factors and seem to be standart. For line spacing both make sense, although relative offsets might be a little more intuitive since users know it from word (1.5, double line spacing etc.).

Overall I think offsets offer more precise control. Factors make it easier to specify double line spacing. Although it is not that complicated with offsets `text.setLineSpacing(font.getLineSpacing(text.getCharacterSize()))`.

Scrolling through the ticket there is an equal amount of people voting for either way. So I don't really know how we find a decision...

Bromeon commented:
I still find this argument important:

> What we also should consider: when the character size increases, the absolute offset will remain the same, making the characters appear closer/tighter. Using a spacing factor on the other hand will preserve the appearance as the characters are varied in size. It probably depends on concrete UI needs, but I find the latter behavior more intuitive.

I'd say users expect that changing the character size effectively scales the text, because that's what most if not all WYSIWYG editors and office software do.

eXpl0it3r commented:
Bump

eXpl0it3r commented:
This PR needs rebasing.

Additionally I've so far only heard votes for using relative spacing, as such I'd say this PR should be adjusted accordingly unless, someone else wants to bring in their opinion and arguments.

eXpl0it3r commented:
This PR needs rebasing.

Additionally I've so far only heard votes for using relative spacing, as such I'd say this PR should be adjusted accordingly unless, someone else wants to bring in their opinion and arguments.

oomek commented:
Is this pull request still alive or has been abandoned?

eXpl0it3r commented:
There was hope that @Foaly would update the PR with the latest discussions, but if this doesn't occur soon-ish, someone from the team will take over his contributions and adapt it for our needs.

Foaly commented:
Found some time to update the code to the "discussion".
I my opinion offsets still make more sense, offer a finer control and are more standard. From how I read this thread there was a equal amount of people for both options. But really I don't care anymore. It's been over 2 years! And I just want to see this functionality be implemented and have this PR merged.

oomek commented:
Great news! But here is a question. Since the letter spacing was mainly needed to reduce the distance between the letters will it accept negative values? Won't it screw the kerning?

Foaly commented:
Right now it accepts negative factors as well. If you use factors in the range [0, 1] letters will move closer together. If you use `-1.f` you can reverse the order of the letters. But as you pointed out it doesn't look too great, because of the kerning.

MarioLiebisch commented:
Maybe add an option to enable/disable kerning? There might be use cases for overlapping characters (e.g. to get a character circled), while not using Unicode control codes. Are those even considered by Freetype? Never tried.

MarioLiebisch commented:
Considering we're not talking about real-time calculations etc. wouldn't it even be better to offer both, offset and scaling? `set*Scaling(…)`, `set*Offset(…)`? I could imagine use cases for both.

Foaly commented:
In my opinion kerning should **always** be enabled as it is an integral part of the font that makes it more readable.
I think we should even prohibit using factors below 0 as it works for reversing but looks bad because it messes with the kerning (maybe an assertion is also enough). If you want to reverse a text you should reverse the used string. I have no idea how well freetype and the current setup handles unicode/overlapping characters.

I think the idea was to keep the API simple hence only one getter/setter. But since I would also say there is use cases for both, adding both would also make sense for me. Maybe others can say what they think about it.

oomek commented:
Has it been merged yet? It's been a while and I'm not sure if it's still maintained or relevant to the current master.

LaurentGomila commented:
I don't know. @Foaly, @eXpl0it3r?

Foaly commented:
I don't know man... At some point I just kind of gave up. This PR has been mostly laying around for 2 and a half years (!!!) for something as ridiculous as adding a factor (or offset...)

Typography is something that I am passionate about. That is why I initially wanted to add this tiny, but fundamental functionality to SFMLs type setting capabilities. I did some proper research, looked into other graphics librarys, css and office software to get an overview to what people are used to, though about what makes sense in SFMLs case and then created an implementation based on it.

It took about half a year for everybody to chip in their code style changes, do some small refactoring and for me to realize that here refactoring is best done in a seperate pull request or not at all. At that point the PR was basically ready to merge.

Then the "discussion" started whether offsets or factors should be used. Some members of the team expressed the opinion that factors would make spacing behave more like people expected. I had the impression that there were some misconceptions about how the spacing worked. I tried to clear those up and noted again that offsets offer more control (which I find desireable for a niche feature like this).
Form my perspective there was about an equal amount of people for either version after all the arguments where brough up. None of the arguments for factors convinced me enough to change the implementation myself. Also after bringing up some counter arguments why the current proposed implementation is not mergeworthy there was radio silence from the SFML team. No more attempts to find a solution and therefor no decision was made.
Especially the lack of leadership or interest in finding a solution left me pretty frustrated.

So I left it and waited. And then after a year there was the statement that this PR was out of date and someone else will take it over if I don't update it to the results of the discussion. I my opinion people only stated their view on the two different possible implementations. Nobody for the SFML team ever made a decision. Noone ever said please implement it this way, we won't merge it otherwise. 

I really did not feel like having this discussion anymore and most team members seemed to prefer factors. So I changed the implementation accordingly. That was 3 month ago.
And then another team member discovered that offsets might be usefull afterall and why not add both factors and offsets. I was way past the irony at that point and in the interest of getting this feature merged I simply said lets go I am all for it, but somebody will probably scream API bloat, so lets hear what people think.

I guess you all know what happend. Sweet sweet silence. No comment from anybody from the team and no long overdue decision.

tl;dr:
In my opinion this PR is a perfect example of everything that is wrong with SFMLs development. There are ridiculously timecomsuming discussions about small API details interleaved with long pauses. Which stalls the development and prohibits progress. And mind you this PR is about adding a getter and a setter for 2 factors/offsets.

I know most of this conservertive behaviour comes from the intention to have a stable API that changes as little as possible. But I am asking in all seriousness: What good is an API that is stable, but severly outdated and has horribly slow development cycles?

I am somebody who has been with this community for years. This development "flow" is activly prohibiting people from contributing. I can't imagine what it must be like for first time contributors. Then again maybe they have not been frustrated this much, when tring to submit a PR to this project.

To answer the question. Yes I still do maintain this PR. I don't really feel like participating in the API discussion anymore. If someone could finally make a decision on how to procced that would be great. I'll be more than happy to implement it.

LaurentGomila commented:
I understand your frustration. And as a member of the team, I am really sorry for that, especially given the hard work of research and implementation that you've put in this PR.

As you figured out, the main problem here is the lack of leadership: there's no one to make the "final decision", everybody in the team and community can share its point of view. This is a good thing, but with the downside that things can get blocked forever when no concensus is found -- which is clearly the case here.

I like factors, but I also like the fact that offsets seem to be more "standard". However since the latest implementation was changed to use factors, let's keep them -- we can still change that later (not in a near future though 😄).

I reviewed the code and it looks fine to me, so as far as I'm concerned this can be merged.

With @eXpl0it3r's agreement, let's put a deadline in seven days: if nobody has anything against it, let's merge it then.

oomek commented:
Photoshop and many similar programs use factors in form of -+% of the space between 2 adjacent letters, and It’s font size agnostic. Some letter pairs have bigger and some smaller gaps, so using offsets in pixels would screw the kerning in my opinion. I vote Factors.

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

eXpl0it3r commented on src/SFML/Graphics/Text.cpp:
```diff
@@ -74,35 +74,39 @@ namespace sf
+m_font               (NULL),
+m_characterSize      (30),
+m_letterSpacingFactor(1.f),
+m_lineSpacingFactor  (0.f),
```
This should be `1.f`.

eXpl0it3r commented on src/SFML/Graphics/Text.cpp:
```diff
@@ -74,35 +74,39 @@ namespace sf
+m_font               (&font),
+m_characterSize      (characterSize),
+m_letterSpacingFactor(1.f),
+m_lineSpacingFactor  (0.f),
```
This should be `1.f`.

Foaly commented:
I've updated the commit.

Foaly commented:
The build system seems to be stuck. It still hasn't built yet. Other than that can we please finally merge this!

eXpl0it3r commented:
SFMLBuildMaster: Build this please.

eXpl0it3r commented:
@Foaly Merging your other font/text fix created some conflicts, would be nice if you could resolve them, sorry. 🙂

Foaly commented:
Fixed!

Foaly commented:
Oh my god! That I still live to see it merged! The emotions!

oomek commented:
This is very exciting news. Thank you.

oomek commented:
Hmmmm, something is not right here.

![sfml letter spacing](https://user-images.githubusercontent.com/2974860/35770511-71e381fe-0914-11e8-92e8-f7006909f573.png)

It seems like the factor 2.0 (200% in theory) multiplies the character width + whitespace instead of just only the whitespace.

oomek commented on src/SFML/Graphics/Text.cpp:
```diff
@@ -425,8 +425,8 @@ void Text::ensureGeometryUpdate() const
     // Precompute the variables needed by the algorithm
-    float whitespaceWidth = m_font->getGlyph(L' ', m_characterSize, isBold).advance + m_letterSpacing;
-    float lineSpacing     = m_font->getLineSpacing(m_characterSize) + m_lineSpacing;
+    float whitespaceWidth = m_font->getGlyph(L' ', m_characterSize, isBold).advance * m_letterSpacingFactor;
```
~~`float whitespaceWidth = m_font->getGlyph(L' ', m_characterSize, isBold).advance * m_letterSpacingFactor;`~~
`float whitespaceWidth = m_font->getGlyph(L' ', m_characterSize, isBold).advance;`
`float letterSpacing = ( whitespaceWidth / 3.0 ) * ( m_letterSpacingFactor - 1.0 );`
`whiteSpaceWidth += letterSpacing;`

oomek commented on src/SFML/Graphics/Text.cpp:
```diff
@@ -525,7 +525,7 @@ void Text::ensureGeometryUpdate() const
 
         // Advance to the next character
-        x += glyph.advance + m_letterSpacing;
+        x += glyph.advance * m_letterSpacingFactor;
```
~~`x += glyph.advance * m_letterSpacingFactor;`~~
`x += glyph.advance * letterSpacing;`

oomek commented on src/SFML/Graphics/Text.cpp:
```diff
@@ -317,8 +317,8 @@ Vector2f Text::findCharacterPos(std::size_t index) const
     bool  isBold          = m_style & Bold;
-    float whitespaceWidth = m_font->getGlyph(L' ', m_characterSize, isBold).advance + m_letterSpacing;
-    float lineSpacing     = m_font->getLineSpacing(m_characterSize)+ m_lineSpacing;
+    float whitespaceWidth = m_font->getGlyph(L' ', m_characterSize, isBold).advance * m_letterSpacingFactor;
```
~~`float whitespaceWidth = m_font->getGlyph(L' ', m_characterSize, isBold).advance * m_letterSpacingFactor;`~~
`float whitespaceWidth = m_font->getGlyph(L' ', m_characterSize, isBold).advance;`
`float letterSpacing = ( whitespaceWidth / 3.0 ) * ( m_letterSpacingFactor - 1.0 );`
`whiteSpaceWidth += letterSpacing;`

oomek replied:
Clearance between characters is usualy 1/3 of the whitespace, so:
for m_letterSpacingFactor = 0.0 the spacing will be -1/3 of a whitespaceWidth (equiv. to -100 in Photoshop )
for m_letterSpacingFactor = 1.0 the spacing will be 1/3 of a whitespaceWidth (equiv. to  0 in Photoshop )
for m_letterSpacingFactor = 2.0 the spacing will be 1/3 of a whitespaceWidth (equiv. to +100 in Photoshop )
for m_letterSpacingFactor = 3.0 the spacing will be 2/3 of a whitespaceWidth (equiv. to +200 in Photoshop )

LaurentGomila commented:
This is a big problem.

Indeed, we should apply the spacing factor to the empty space between characters and not to the global advance. I'm not sure if those arbitrary calculations (whitespace_width / 3) are "correct"; I even wonder if for spacingFactor == 1.0 it produces the exact same result as before this PR.

Can't we use some extra Glyph information? For example, subtracting the glyph's width to its advance should in theory give the empty space. And we shouldn't forget about kerning, too.

oomek commented:
I seriously doubt we can extract the type width from the font as I looked at some fonts in the editor and the width already contains the clearance between the characters. My math is the closest approximation you can get. For the large font like 60pt the whitespace = 25 pixels, the clearance between the chracters = 9 pixels:
`letterSpacing = ( whitespaceWidth / 3.0 ) * ( m_letterSpacingFactor - 1.0 )`
so for m_letterSpacingFactor == 1.0
`letterSpacing = ( 25.0 / 3.0 ) * ( 1.0 - 1.0 ) = 0.0`
for m_letterSpacingFactor == 2.0
`letterSpacing = ( 25.0 / 3.0 ) * ( 2.0 - 1.0 ) = 8.333`
for m_letterSpacingFactor == 3.0
`letterSpacing = ( 25.0 / 3.0 ) * ( 3.0 - 1.0 ) = 16.666`
for m_letterSpacingFactor == 0.0
`letterSpacing = ( 25.0 / 3.0 ) * ( 0.0 - 1.0 ) = -8.333`

If the getLineSpacing function returns float we should be fine, but if it's an int we are screwed for small characters. That's why I would need this compiled to make sure I'm right.

Foaly commented:
Thank you for testing and reporting! Indeed there seems to be an error in the implementation.
Right now the letterSpacingFactor is applied only to the advance of a glyph. The actual distance between two glyphs is put together from the advance of the first glyph and the kerning offset between the second and first. They are applied at different points, so I must have missed it.
I would like to try taking the kerning into account first, before applying some scaling. I would be interested in how you came up with `/ 3.0`?

oomek commented:
I've rendered some fonts in Photoshp in large scale and counted pixels and checked how the clearance is behaving for different letter spacings. it was almost always 1/3 of a space. I dod not count the offset between the characters as this changes from character to character, but rather the delta between different letter spacing. Delta was the same for each glyph pairs.

oomek commented:
If .advance returns the distance with kerning applied that's even better as according to my tests, the delta in pixels is the same for each letter pair when you set +100 or +200 tracking

LaurentGomila commented:
> I seriously doubt we can extract the type width from the font as I looked at some fonts in the editor and the width already contains the clearance between the characters.

I don't know what editor you're talking about, but in SFML, `Glyph::advance` is the total distance to the next glyph, and `Glyph::bounds::width` is the exact width of the glyph's pixels. So the difference between those *should be exactly* what we are looking for. I'd definitely give it a try as a fix for this PR.

> If .advance returns the distance with kerning applied

It doesn't, and it can't since kerning depends not only on the character but also on the next one (in other words, it is defined for each possible pair of characters). Kerning is retrieved with a dedicated function from sf::Font.

oomek commented:
It's FontForge.
Glyph::bounds::width will return from what I understand the distance between the vertical metric lines, which is definitely not something we are looking for as it contains the clearance between characters.
I'm attaching a screenshot
![fontforge](https://user-images.githubusercontent.com/2974860/35781091-4db7b9a2-09dd-11e8-90ce-aef8fa9516b8.png)

LaurentGomila commented:
> Glyph::bounds::width will return from what I understand the distance between the vertical metric lines, which is definitely not something we are looking for as it contains the clearance between characters.

No it does not. Again, it's defined as the width of the glyph's bitmap, so there's precisely no extra space in it. It's not taken from the font's metrics, just deduced from the bitmap.

However I just tested my idea, and it doesn't work. The "extra space" is not what we want to use, as some glyphs will have a big one and some others none at all, leading to inconsistent results within the same text. Moreover, it doesn't correctly applies to the space character, which has only extra space.

And regardless of the chosen solution, I'm not sure what to do with the kerning: it's a negative offset, which means that it brings glyphs closer to each other. Applying the factor to it would make the glyphs even more close, not sure this is what we want.

oomek commented:
Could you try my method just to be sure. If it's close enough I'm fine with it as long as it does not produce uneven spacings between characters.

LaurentGomila commented:
I'll try to find how other libraries implement it, and I'll give your solution a try if I get some time to test it.

oomek commented:
Thanks

LaurentGomila commented:
Qt have both options, and the relative spacing is implemented exactly as in this PR: they multiply the whole advance by the factor; which leads to the same weird results. This is really disappointing.

Others seem to use absolute offsets rather than relative factors -- maybe this is the reason :(

oomek commented:
I've built SFML with my changes and the result is better than I anticipated. 
Vide below:
https://www.youtube.com/watch?v=K5ISsAalO6k

LaurentGomila commented:
This is indeed really nice :)

You should also test interaction with kerning (add "AV" to your text, for example), and tiny character sizes (below 10).

LaurentGomila commented:
After thinking more about it, I think it is a good idea to base the spacing on something that is specific to the font and character size, but remains the same for each glyph. It's not accurate, because strictly speaking, applying a factor should modify the existing spacing and not some artificial one, but as we've seen, this is what produces the best looking results.

As an additional test, you can try with a monospace font, just to make sure it still looks good with those fonts.

If your modification successfully passes all these tests, then I'm ok for applying it. But I'd like to get feedback from other members (including @Foaly and other people involved in this PR), if possible.

oomek commented:
The spacing of small fonts and AV pairs looks also fine, except one thing. Smal fonts are a bit blurry when I adjust the scaling factor, but that’s ro be expected as the position is no longer snapped to pixels and small fonts position is optimizd to look crisp. I’m experimenting now to see if it can be overcome by quantizing the offsets.

Foaly commented:
Oh the irony! Excuse me but I think this is hilarious!
I haven't gotten around to do some test myself, but from what I read from your descriptions, I see two ways of fixing this.
I guess the proper solution would be to somehow™ get the actual "whitespace" between two characters (meaning the distance between rightmost black pixel of the left character and the leftmost black pixel of the right character) containing kerning and everything and then apply the factor to that. 
As @Laurent explained this is not easily achievable, but maybe there is a way noone here has thought about yet. This would have the nice feature that the spacing is font and character size dependent. Since this seems hard to do a somewhat "dirty" fix would be to use magic numbers and ratios to achieve similar looking results (as @oomek suggested). I'm not a fan of such a design and would prefer a mathematically sane calculation, but I guess if there is no cleaner way I would be fine with it.

The second, simpler solution is to simply add an offset, that is added between two characters. For historic reasons I am gonna favour that one.

oomek commented:
I do not see any irony here. Why would you prefer a fixed offset that is not dependant on the font size?

LaurentGomila commented:
> As @laurent explained this is not easily achievable

It is. I tested it, and it looks terrible. Extra spacing between characters must not depend on the actual one, but rather on some fixed value which only depends on character size.

Foaly commented:
Well it is ironic, because we spend about two years discussing whether to use factors or offsets. And now it turns out factors have this major flaw. But it was more of a design discussion and appearently this was overlooked. So I guess I also have to appologize for not testing thoroughly enough.
That is why I said I favour offset. Also it is the simplest solution to the problme (as in similar result, but no magic numbers).
But I like @LaurentGomila explained the proper solution would be to have something character size dependent, which would be fine by me. So thinking about it again maybe using the `whitespaceWidth` isn't such a "magic numbers" solution. Could you post a link to branch with your changes?

oomek commented:
SegoeUI
https://www.youtube.com/watch?v=SYVYpDhi_AE
Consolas Monospaced
https://www.youtube.com/watch?v=N6kshXQ9rUU

Forgive me for my lack of professionalism, but I did not fork SFML, I didn't want to add a duplicate of your commit with just some minor changes. would it be ok to just upload the text.cpp alone and post here the link?

Foaly commented:
Sure! Or maybe just the diff.

oomek commented:
https://github.com/oomek/SFML/commit/1e24d118d8776caf3381fd5e80b78317a21570cb

oomek commented:
Not sure still If I have to ammend the lineSpacing as well. Will look into it when my Fibromyalgia flare goes away.

LaurentGomila commented:
Line spacing should be ok.

Foaly commented:
Looks fine by me. I'd say you could send a pull request for it.
The line spacing works much simpler and should be fine, but of course feel free to take a look :)
